In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split

In [3]:
df = pd.read_csv("titanic_modified.csv") 
df.head(20)

pclass  survived                                               name  \
0        1         1                      Allen, Miss. Elisabeth Walton   
1        1         1                     Allison, Master. Hudson Trevor   
2        1         0                       Allison, Miss. Helen Loraine   
3        1         0               Allison, Mr. Hudson Joshua Creighton   
4        1         0    Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   
5        1         1                                Anderson, Mr. Harry   
6        1         1                  Andrews, Miss. Kornelia Theodosia   
7        1         0                             Andrews, Mr. Thomas Jr   
8        1         1      Appleton, Mrs. Edward Dale (Charlotte Lamson)   
9        1         0                            Artagaveytia, Mr. Ramon   
10       1         0                             Astor, Col. John Jacob   
11       1         1  Astor, Mrs. John Jacob (Madeleine Talmadge Force)   
12       1         1                      Aubart, Mme. Leontine Pauline   
13       1         1                       Barber, Miss. Ellen 'Nellie'   
14       1         1               Barkworth, Mr. Algernon Henry Wilson   
15       1         0                                Baumann, Mr. John D   
16       1         0                           Baxter, Mr. Quigg Edmond   
17       1         1    Baxter, Mrs. James (Helene DeLaudeniere Chaput)   
18       1         1                              Bazzani, Miss. Albina   
19       1         0                               Beattie, Mr. Thomson   

       sex      age  sibsp  parch    ticket      fare    cabin embarked boat  \
0   female  29.0000      0      0     24160  211.3375       B5        S    2   
1     male   0.9167      1      2    113781  151.5500  C22 C26        S   11   
2   female   2.0000      1      2    113781  151.5500  C22 C26        S  NaN   
3     male  30.0000      1      2    113781  151.5500  C22 C26        S  NaN   
4   female  25.0000      1      2    113781  151.5500  C22 C26        S  NaN   
5     male  48.0000      0      0     19952   26.5500      E12        S    3   
6   female  63.0000      1      0     13502   77.9583       D7        S   10   
7     male  39.0000      0      0    112050    0.0000      A36        S  NaN   
8   female  53.0000      2      0     11769   51.4792     C101        S    D   
9     male  71.0000      0      0  PC 17609   49.5042      NaN        C  NaN   
10    male  47.0000      1      0  PC 17757  227.5250  C62 C64        C  NaN   
11  female  18.0000      1      0  PC 17757  227.5250  C62 C64        C    4   
12  female  24.0000      0      0  PC 17477   69.3000      B35        C    9   
13  female  26.0000      0      0     19877   78.8500      NaN        S    6   
14    male  80.0000      0      0     27042   30.0000      A23        S    B   
15    male      NaN      0      0  PC 17318   25.9250      NaN        S  NaN   
16    male  24.0000      0      1  PC 17558  247.5208  B58 B60        C  NaN   
17  female  50.0000      0      1  PC 17558  247.5208  B58 B60        C    6   
18  female  32.0000      0      0     11813   76.2917      D15        C    8   
19    male  36.0000      0      0     13050   75.2417       C6        C    A   

     body                        home.dest cabinReduced  
0     NaN                     St Louis, MO            B  
1     NaN  Montreal, PQ / Chesterville, ON            C  
2     NaN  Montreal, PQ / Chesterville, ON            C  
3   135.0  Montreal, PQ / Chesterville, ON            C  
4     NaN  Montreal, PQ / Chesterville, ON            C  
5     NaN                     New York, NY            E  
6     NaN                       Hudson, NY            D  
7     NaN                      Belfast, NI            A  
8     NaN              Bayside, Queens, NY            C  
9    22.0              Montevideo, Uruguay          NaN  
10  124.0                     New York, NY            C  
11    NaN                     New York, NY         

### Zapoznanie się z funkcja  train_test_split() z biblioteki Scikit-learn
**sklearn.model_selection.train_test_split(arrays, test_size=None, train_size=None, random_state=None, shuffle=True, stratify=None)**

Funkcja **train_test_split()** służy do podziału danych na zbiór treningowy i testowy, zgodnie z podanymi przez nas parametrami. Przyjmuje jako argumenty ramki danych (DataFrame), tablice NumPy, listy lub macierze Scipy, które mają zostać rozdzielone na te dwa zbiory. Możemy określić rozmiar zbioru testowego lub treningowego — nie są to wartości z góry narzucone, co daje możliwość indywidualnego dostosowania proporcji podziału do konkretnego problemu.  

Parametr **random_state** pozwala uzyskać powtarzalność losowego podziału, **shuffle** decyduje o tym, czy dane mają zostać przemieszane przed podziałem, a **stratify** umożliwia zachowanie proporcji wybranej zmiennej w obu zbiorach. Dzięki temu możemy uniknąć sytuacji, w której pewne klasy są obecne tylko w zbiorze treningowym, a brakuje ich w testowym.

####  Tworzenie zmiennej z listą trzech kolumn: cabin, cabinReduced, sex

In [6]:
col_name = ["cabin", "cabinReduced", "sex"]
df[col_name].head(20)

cabin cabinReduced     sex
0        B5            B  female
1   C22 C26            C    male
2   C22 C26            C  female
3   C22 C26            C    male
4   C22 C26            C  female
5       E12            E    male
6        D7            D  female
7       A36            A    male
8      C101            C  female
9       NaN          NaN    male
10  C62 C64            C    male
11  C62 C64            C  female
12      B35            B  female
13      NaN          NaN  female
14      A23            A    male
15      NaN          NaN    male
16  B58 B60            B    male
17  B58 B60            B  female
18      D15            D  female
19       C6            C    male

#### Dzielenie znioru na treningowy i testowy

In [9]:
X = df[col_name]
y = df["survived"]

X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.3, random_state=0)

print(f"Rozmiar X_train {X_train.shape}, X_test {X_test.shape}, y_train {y_train.shape}, y_test {y_test.shape}")


Rozmiar X_train (916, 3), X_test (393, 3), y_train (916,), y_test (393,)


Zmienna X zawiera 3 kolumny: **kabina**, **zredukowana kabina**, **płeć**.    
Zmienna y to nasza kolumna **survived**, czyli informacja, czy nasza osoba przeżyła (zmienna zależna).   
Podział danych został ustawiony 70% treningowych i 30% testowych, X_train i y_train mają 916 próbek, a X_test i y_test mają 393 próbki. 



#### Badanie rozkładu etykiet

In [12]:
for zmienna in col_name:
  unique_test = [x for x in X_test[zmienna].unique() if x not in X_train[zmienna].unique()]
  print(f"{zmienna} ma {len(unique_test)} etykiet, które występują się w zbiorze testowym, a nie pojawiają się w treningowym.")
  print(unique_test)

cabin ma 37 etykiet, które występują się w zbiorze testowym, a nie pojawiają się w treningowym.
[nan, 'E12', 'C104', 'A31', 'D11', 'D48', 'D10 D12', 'B38', 'D45', 'C50', 'C31', 'B82 B84', 'A32', 'C53', 'B10', 'C70', 'A23', 'C106', 'C46', 'E58', 'B11', 'F E69', 'B80', 'E39 E41', 'D22', 'E40', 'A19', 'C32', 'B79', 'C45', 'B22', 'B39', 'C47', 'B101', 'A7', 'E52', 'F38']
cabinReduced ma 1 etykiet, które występują się w zbiorze testowym, a nie pojawiają się w treningowym.
[nan]
sex ma 0 etykiet, które występują się w zbiorze testowym, a nie pojawiają się w treningowym.
[]


W przypadku zmiennej **sex** obie etykiety występują zarówno w jednym, jak i w drugim zbiorze.  
Jeżeli chodzi o zmienną **cabin** to pojawia się 37 etykiet, które występują w zbiorze testowym, a nie ma ich w zbiorze treningowym, to znaczy, że model uczony na zbiorze treningowym nie miał okazji zobaczyć tych wartości *cabin* wcześniej, jest to zła informacja.  
W kwestii zmiennej **cabin_Reduced** tylko jedna unikalna etykieta (nan) występuje w zbiorze testowym, a nie w treningowym, wynika to z tego, że nan nie jest równe nan.

#### Kodowanie zmiennych kategorycznych do zmiennych liczbowych

In [13]:
# Lista słowników do kodowania etykiet
encoding_dicts = []

# Dla każdej kolumny kategorycznej tworzymy słownik z przypisanymi wartościami liczbowymi
for column in col_name:
    label_to_int = {}
    for idx, label in enumerate(df[column].unique()):
        label_to_int[label] = idx
    encoding_dicts.append(label_to_int)

# Tworzenie nowych zakodowanych kolumn w zbiorach X_train i X_test
for idx in range(len(col_name)):
    col = col_name[idx]
    mapping = encoding_dicts[idx]
    
    X_train[f"{col}_encoded"] = X_train[col].map(mapping)
    X_test[f"{col}_encoded"] = X_test[col].map(mapping)

# Wyświetlenie przetworzonych zbiorów
display(X_train)
display(X_test)


cabin cabinReduced     sex  cabin_encoded  cabinReduced_encoded  \
501    NaN          NaN  female              6                     5   
588    NaN          NaN  female              6                     5   
402    NaN          NaN  female              6                     5   
1193   NaN          NaN    male              6                     5   
686    NaN          NaN  female              6                     5   
...    ...          ...     ...            ...                   ...   
763    NaN          NaN  female              6                     5   
835    NaN          NaN    male              6                     5   
1216   NaN          NaN  female              6                     5   
559    NaN          NaN  female              6                     5   
684    NaN          NaN  female              6                     5   

      sex_encoded  
501             0  
588             0  
402             0  
1193            1  
686             0  
...           ...  
763             0  
835             1  
1216            0  
559             0  
684             0  

[916 rows x 6 columns]

cabin cabinReduced     sex  cabin_encoded  cabinReduced_encoded  \
1139   NaN          NaN    male              6                     5   
533    NaN          NaN  female              6                     5   
459    NaN          NaN    male              6                     5   
1150   NaN          NaN    male              6                     5   
393    NaN          NaN    male              6                     5   
...    ...          ...     ...            ...                   ...   
914    NaN          NaN    male              6                     5   
580    NaN          NaN  female              6                     5   
1080   NaN          NaN    male              6                     5   
1249   F38            F    male            186                     7   
88     NaN          NaN  female              6                     5   

      sex_encoded  
1139            1  
533             0  
459             1  
1150            1  
393             1  
...           ...  
914             1  
580             0  
1080            1  
1249            1  
88              0  

[393 rows x 6 columns]

#### Sprawdzanie liczby brakujących wartości w zmodyfikowanych zbiorach:

In [17]:
for zmienna in col_name:
  print(f"Liczba NaN w kolumnie {zmienna}_map w zbiorze X_train: {X_train[zmienna+'_encoded'].isnull().sum()}, w X_test: {X_test[zmienna+'_encoded'].isnull().sum()}")

Liczba NaN w kolumnie cabin_map w zbiorze X_train: 0, w X_test: 0
Liczba NaN w kolumnie cabinReduced_map w zbiorze X_train: 0, w X_test: 0
Liczba NaN w kolumnie sex_map w zbiorze X_train: 0, w X_test: 0


#### Komentarz na temat powyższych wyników:
Brak wartości NaN w kodowanych kolumnach świadczy o tym, że mapowanie przebiegło pomyślnie.   
Nie wystąpiły żadne nowe, niespodziewane wartości w zbiorze testowym, które nie byłyby znane w treningowym.   
Kodowanie zostało przeprowadzone przy użyciu słowników utworzonych na podstawie pełnego zbioru danych, co zapewniło, że etykiety z X_test również znalazły swoje odpowiedniki liczbowo zakodowane. Dzięki temu uniknięto powstawania NaN po .map()





#### Zastąpienie brakujących wartości liczbą 0

In [18]:
encoded_cols = [col + '_encoded' for col in col_name]

X_train[encoded_cols] = X_train[encoded_cols].fillna(0)
X_test[encoded_cols] = X_test[encoded_cols].fillna(0)

Ma to sens w przypadku, gdy **0** jest traktowane jako oddzielna kategoria, czyli kategoria "braków", nasze zmienne cabin, sex i cabin_Reduced są jakościowe, nie ilościowe, gdy stałyby się ilościowe, a **0** byłoby wartością kodującą jakieś dane, to nie mogę tego zrobić.

#### Porównanie wartości unikalnych w zbiorze testowym i treningowym

In [25]:
print(len(X_train.groupby(col_name).size()))
print(len(X_train))
print(len(X_test.groupby(col_name).size()))
print(len(X_test))

178
916
77
393


In [26]:
print(len(X_train.groupby(col_name[1:]).size()))
print(len(X_train))
print(len(X_test.groupby(col_name[1:]).size()))
print(len(X_test))

14
916
12
393


#### Przed redukcją **cabin**
W zbiorze treningowym mamy 178 unikalnych kombinacji w kolumnach: **cabin, cabin_Reduced i sex**, co oznacza, że niektóre kombinacje pojawiają się wielokrotnie, ponieważ w zbiorze mamy 916 wierszy, wynika z tego, że większość cech w zbiorze jest powtarzalna.  
W zbiorze testowym mamy 77 unikalnych kombinacji, z tego wynika, że zbiór testowy zawiera mniej różnorodne przypadki niż treningowy zbiór danych. 
Jeśli większość kombinacji w zbiorze testowym jest obecna w zbiorze treningowym, model będzie miał szansę lepiej generalizować na testowy zbiór, ponieważ napotka te same kombinacje cech.
#### Po redukcji **cabin**
W zbiorze treningowym liczba unikalnych kombinacji zmniejszyła się do 14, a w zbiorze testowym do 12. 


Przy mapowaniu liczba etykiet pozostaje taka sama, a po redukcji etykiet jest mniej. Redukcja ma wpływ na działanie modelu i wyniki predykcji, natomiast mapowanie sprawia, że dane stają się łatwiejsze dla modelu. 

Zmniejszenie liczby unikalnych kategorii w cabin sprawia, że model nie musi radzić sobie z dużą liczbą różnych kategorii. To zmniejsza złożoność danych i zapobiega przeuczeniu (overfitting), ponieważ model nie musi zapamiętywać zbyt szczegółowych informacji.

